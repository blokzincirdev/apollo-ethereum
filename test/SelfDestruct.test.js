const gql = require('graphql-tag')

const SelfDestruct = artifacts.require('SelfDestruct')

const config = {
  contracts: {
    SelfDestruct: {
      canSelfDestruct: true,
      fields: {
        integer: 'Int',
        'Log.randomNumber': 'Int',
      },
    },
  },
}

const QUERY = gql`
  query($address: Address) {
    SelfDestruct(address: $address) {
      __typename
      Log {
        message
        randomNumber
        _timestamp
      }
      ... on SelfDestructActive {
        message
        integer
      }
    }
  }
`

const MUTATION_DESTROY = gql`
  mutation($address: Address) {
    SelfDestruct(address: $address) {
      destroy
    }
  }
`

contract('SelfDestruct', () => {
  let client

  before(async () => {
    client = await createClient(config, { SelfDestruct })
  })

  it('generates graphql schema', async () => {
    expect(client.source).to.matchSnapshot()
  })

  it('queries available fields on active contracts', async () => {
    const data = await client.execute(QUERY, { address: SelfDestruct.address })
    const { Log, ...rest } = data.SelfDestruct
    expect(rest).to.deep.equal({
      __typename: 'SelfDestructActive',
      message: 'only available on active contracts',
      integer: 42,
    })
    expect(Log).to.containSubset([
      { message: 'Creating the contract', randomNumber: 20 },
    ])
  })

  it('does not query state on destroyed contracts', async () => {
    await client.execute(MUTATION_DESTROY, { address: SelfDestruct.address })

    const data = await client.execute(QUERY, {
      address: SelfDestruct.address,
    })
    const { Log, ...rest } = data.SelfDestruct
    expect(rest).to.deep.equal({
      __typename: 'SelfDestructComplete',
      // note: no other fields to appear
    })
    expect(Log).to.containSubset([
      { message: 'Creating the contract', randomNumber: 20 },
      { message: 'Destroying the contract', randomNumber: 30 },
    ])
  })
})
