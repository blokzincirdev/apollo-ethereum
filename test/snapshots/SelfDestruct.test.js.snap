// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contract: SelfDestruct generates graphql schema 1`] = `
"directive @erc1820(interfaceName: String) on FIELD_DEFINITION

directive @mappingIndex(mapping: String) on FIELD_DEFINITION

scalar Address

type Mutation {
  SelfDestruct(address: Address!): SelfDestructMutative
}

type Query {
  SelfDestruct(address: Address!): SelfDestruct
}

interface SelfDestruct {
  _address: Address
  Log: [SelfDestruct_Log]
}

type SelfDestruct_Log {
  message: String
  randomNumber: Int
  _timestamp: Timestamp
}

type SelfDestructActive implements SelfDestruct {
  _address: Address
  Log: [SelfDestruct_Log]
  integer: Int
  message: String
}

type SelfDestructComplete implements SelfDestruct {
  _address: Address
  Log: [SelfDestruct_Log]
}

type SelfDestructMutative {
  destroy: Boolean
}

scalar Timestamp
"
`;
